src/vs/workbench/contrib/notebook/browser/services/notebookExecutionServiceImpl.ts

```diff
@@ -102,6 +116,46 @@ export class NotebookExecutionService implements INotebookExecutionService, IDis
 		this.cancelNotebookCellHandles(notebook, Array.from(cells, cell => cell.handle));
 	}
 
+	async formatNotebookCells(cellExecutions: INotebookCellExecution[], notebookModel: INotebookTextModel): Promise<void> {
```

Comment:
I don't think this should go into the execution service itself just because there's no reason for an "execution service" to know about formatters. You can have something else that listens to events on the service and formats them. Assuming you want to do this before the execution and block execution on formatting, we would need a new API on the service to do this. Not sure if this is like what @rebornix had in mind but I'm imagining something like this

```
interface IKernelExecutionParticipant {
    onWillExecuteCell(execution: INotebookCellExecution): Promise<void>;
}

// and on the service
export interface INotebookExecutionStateService {
    registerExecutionParticipant(participant: IKernelExecutionParticipant)
}
```

then you can call those participants before doing the execution and await them. This is different from the existing event because listening to the event won't block the service on any work the listener wants to do.

Reply:
@roblourens we were reluctant to introduce the participant concept which we used to have in Jupyter and later on became difficult to manage. 

With that said, it makes sense to me to have `IKernelExecutionParticipant` and we contribute a participant from `contrib/formatting`.